src/detail/fastrtps_dynamic_data.cpp

```diff
@@ -36,285 +38,418 @@
-    fastrtps__size_t_to_uint32_t(id));
   if (!loaned_data_impl_handle) {
+    RCUTILS_SET_ERROR_MSG("Could not init new data");
+    return RCUTILS_RET_ERROR;
+  }
+
+  loaned_data_impl_handle->ref_type =
+    data_handle->ref_type->loan_value(fastdds__size_t_to_uint32_t(id));
+  if (!loaned_data_impl_handle->ref_type) {
     RCUTILS_SET_ERROR_MSG("Could not loan dynamic data");
+    delete loaned_data_impl_handle;
     return RCUTILS_RET_ERROR;
   }
 
   loaned_data_impl->handle = std::move(loaned_data_impl_handle);
   return RCUTILS_RET_OK;
 }
 
-
-rcutils_ret_t
-fastrtps__dynamic_data_return_loaned_value(
+rcutils_ret_t fastdds__dynamic_data_return_loaned_value(
   rosidl_dynamic_typesupport_serialization_support_impl_t * serialization_support_impl,
   rosidl_dynamic_typesupport_dynamic_data_impl_t * data_impl,
   const rosidl_dynamic_typesupport_dynamic_data_impl_t * inner_data_impl)
 {
-  (void) serialization_support_impl;
-  FASTRTPS_CHECK_RET_FOR_NOT_OK_WITH_MSG(
-    static_cast<DynamicData *>(data_impl->handle)
-    ->return_loaned_value(static_cast<const DynamicData *>(inner_data_impl->handle)),
-    "Could not return loaned value"
-  );
+  (void)serialization_support_impl;
+
+  auto data_handle =
+    static_cast<const fastdds__rosidl_dynamic_typesupport_dynamic_data_impl *>(data_impl->handle);
+  if (!data_handle || !data_handle->ref_type) {
+    RCUTILS_SET_ERROR_MSG("Could not get handle to data impl");
+    return RCUTILS_RET_INVALID_ARGUMENT;
+  }
+
+  auto inner_data_handle =
+    static_cast<const fastdds__rosidl_dynamic_typesupport_dynamic_data_impl *>(inner_data_impl->
+    handle);
+  if (!inner_data_handle || !inner_data_handle->ref_type) {
+    RCUTILS_SET_ERROR_MSG("Could not get handle to inner data impl");
+    return RCUTILS_RET_INVALID_ARGUMENT;
+  }
+
+  FASTDDS_CHECK_RET_FOR_NOT_OK_WITH_MSG(
+    data_handle->ref_type->return_loaned_value(inner_data_handle->ref_type),
+    "Could not return loaned value");
+  //TODO(richiware) remove structure?
   return RCUTILS_RET_OK;
 }
 
-
 rcutils_ret_t
-fastrtps__dynamic_data_get_name(
+fastdds__dynamic_data_get_name(
   rosidl_dynamic_typesupport_serialization_support_impl_t * serialization_support_impl,
   const rosidl_dynamic_typesupport_dynamic_data_impl_t * data_impl,
   const char ** name,
   size_t * name_length)
 {
-  (void) serialization_support_impl;
-  std::string tmp_name = static_cast<DynamicData *>(data_impl->handle)->get_name();
+  (void)serialization_support_impl;
+
+  auto data_handle =
+    static_cast<const fastdds__rosidl_dynamic_typesupport_dynamic_data_impl *>(data_impl->handle);
+  if (!data_handle || !data_handle->ref_type) {
+    RCUTILS_SET_ERROR_MSG("Could not get handle to data impl");
+    return RCUTILS_RET_INVALID_ARGUMENT;
+  }
+
+  std::string tmp_name = data_handle->ref_type->type()->get_name().to_string();
   *name = rcutils_strdup(tmp_name.c_str(), rcutils_get_default_allocator());
```

Comment:
I know you didn't add this, but this should check for a NULL return value:

```suggestion
  *name = rcutils_strdup(tmp_name.c_str(), rcutils_get_default_allocator());
  if (*name == nullptr) {
    RCUTILS_SET_ERROR_MSG("Failed to duplicate name for data name");
    return RCUTILS_RET_BAD_ALLOC;
  }
```

Reply:
Done in 2ec21b3